homeassistant/components/backup/manager.py

```diff
@@ -244,12 +244,21 @@ async def _async_upload_backup(
     ) -> None:
         """Upload a backup to selected agents."""
         LOGGER.warning("Uploading backup %s to agents %s", backup.backup_id, agent_ids)
+
+        async def send_backup() -> AsyncGenerator[bytes]:
+            with await self.hass.async_add_executor_job(path.open, "rb") as f:
+                while chunk := f.read(2**20):
+                    yield chunk
+
+        async def open_backup() -> AsyncGenerator[bytes]:
+            return send_backup()
+
         self.syncing = True
         try:
             sync_backup_results = await asyncio.gather(
                 *(
                     self.backup_agents[agent_id].async_upload_backup(
-                        path=path,
+                        open_stream=open_backup,
```

Comment:
Why do we need to pass a function that returns a coroutine that returns the generator instead of just passing the generator function? Ie, I'd naively pass `send_backup` directly.

Reply:
Does the http request need to happen after `async_upload_backup` is called?

Reply:
Yes, the agent may need to access the data more than once (the cloud agent needs to do this in fact).

Reply:
Ok